File: prog19.cpp
#include <iostream>
using namespace std;

class ListNode
{
public:
    int val;
    ListNode* next;
    ListNode(int x, ListNode* n=NULL)
    {
        val=x;
        next=n;
    }
};

void display(ListNode* temp)
{
    while (temp->next != NULL)
    {
        cout << temp->val << "->";
        temp=temp->next;
    }
    cout << temp->val << endl;
}

ListNode* removeNthFromEnd(ListNode *head, int n)
{
    ListNode *p1,*p2;
    p1=head;p2=p1;
    for (int i=0;i<n;i++)
    {
        p2=p2->next;
    }
    while (p2->next != NULL)
    {
        p1=p1->next;
        p2=p2->next;
    }
    p1->next=p1->next->next;
    return head;
}

int main()
{
    ListNode * n1 = new ListNode(1);
    ListNode * n2 = new ListNode(2);
    ListNode * n3 = new ListNode(3);
    ListNode * n4 = new ListNode(4);
    ListNode * n5 = new ListNode(5);
    n1->next=n2;
    n2->next=n3;
    n3->next=n4;
    n4->next=n5;
    display(n1);
    display(removeNthFromEnd(n1,2));
    return 0;
}

File: prog19.py
class ListNode:
    def __init__(self,x):
        self.val=x
        self.next = None
    def display(self):
        if self.next == None:
            print (self.val)
        else:
            print (self.val,"->",end="")
            self.next.display()

def removeNthFromEnd(head,n):
    p1=head
    p2=p1
    for i in range(n):
        p2=p2.next
    while p2.next != None:
        p1=p1.next
        p2=p2.next
    p1.next=p1.next.next
    return head

n1=ListNode(1)
n2=ListNode(2)
n3=ListNode(3)
n4=ListNode(4)
n5=ListNode(5)
n1.next=n2
n2.next=n3
n3.next=n4
n4.next=n5
n1.display()
removeNthFromEnd(n1,2).display()


